File: Segment-Tree/3525.Find X Value of Array II.cpp
#include <vector>
using namespace std;

struct Node
{
    int count[5] = {0};
    int prod = 0;
};

class SegmentTree
{
public:
    int n;
    int k;
    vector<Node> segTree;

    SegmentTree(vector<int> &nums, int k)
    {
        this->k = k;
        this->n = nums.size();
        segTree.resize(4 * n, Node());
        build(0, 0, n - 1, nums);
    }

    void build(int i, int l, int r, vector<int> &nums)
    {
        if (l == r)
        {
            leafNode(i, nums[l]);
            return;
        }
        int mid = (l + r) >> 1;
        build(2 * i + 1, l, mid, nums);
        build(2 * i + 2, mid + 1, r, nums);

        segTree[i] = mergeNodes(segTree[2 * i + 1], segTree[2 * i + 2]);
    }

    void leafNode(int i, int val)
    {
        for (int x = 0; x < k; x++)
        {
            segTree[i].count[x] = 0;
        }

        int rem = val % k;
        segTree[i].count[rem] = 1;
        segTree[i].prod = rem;
    }

    Node mergeNodes(Node &leftChild, Node &rightChild)
    {
        Node result;
        result.prod = (leftChild.prod * rightChild.prod) % k;

        for (int x = 0; x < k; x++)
        {
            result.count[x] = leftChild.count[x];
        }

        for (int x = 0; x < k; x++)
        {
            int newRemain = (leftChild.prod * x) % k;
            result.count[newRemain] += rightChild.count[x];
        }
        return result;
    }

    void segTreeUpdate(int i, int l, int r, int idx, int val)
    {
        if (l == r)
        {
            leafNode(i, val);
            return;
        }

        int mid = (l + r) >> 1;
        if (idx <= mid)
        {
            segTreeUpdate(2 * i + 1, l, mid, idx, val);
        }
        else
        {
            segTreeUpdate(2 * i + 2, mid + 1, r, idx, val);
        }

        segTree[i] = mergeNodes(segTree[2 * i + 1], segTree[2 * i + 2]);
    }

    void update(int idx, int val) { segTreeUpdate(0, 0, n - 1, idx, val); }

    Node segTreeQuery(int start, int end, int i, int l, int r)
    {
        if (l >= start && r <= end)
        {
            return segTree[i];
        }

        int mid = (l + r) >> 1;

        if (end <= mid)
        {
            return segTreeQuery(start, end, 2 * i + 1, l, mid);
        }

        if (start > mid)
        {
            return segTreeQuery(start, end, 2 * i + 2, mid + 1, r);
        }

        Node left = segTreeQuery(start, end, 2 * i + 1, l, mid);
        Node right = segTreeQuery(start, end, 2 * i + 2, mid + 1, r);
        return mergeNodes(left, right);
    }

    Node query(int start, int end)
    {
        return segTreeQuery(start, end, 0, 0, n - 1);
    }
};

class Solution
{
public:
    vector<int> resultArray(vector<int> &nums, int k,
                            vector<vector<int>> &queries)
    {
        int n = nums.size();
        SegmentTree segTree(nums, k);
        vector<int> result;

        for (auto &q : queries)
        {
            int idx = q[0];
            int val = q[1];
            int start = q[2];
            int x = q[3];

            segTree.update(idx, val);

            Node resultNode = segTree.query(start, n - 1);
            result.push_back(resultNode.count[x]);
        }
        return result;
    }
};

File: Segment-Tree/3525.Find X Value of Array II.ts
interface SegTreeNode {
    prod: number;
    count: number[];
}

class SegTree {
    private n: number;
    private k: number;
    private segTree: SegTreeNode[];
    private nums: number[];
    constructor(n: number, k: number, nums: number[]) {
        this.k = k;
        this.n = n;
        this.nums = nums;
        this.segTree = new Array(4 * n);
        this.build(0, 0, n - 1);
    }

    private build(i: number, l: number, r: number): void {
        if (l == r) {
            this.leafNode(i, this.nums[l]);
            return;
        }

        let mid = (l + r) >> 1;

        this.build(2 * i + 1, l, mid);
        this.build(2 * i + 2, mid + 1, r);
        this.segTree[i] = this.mergeNodes(this.segTree[2 * i + 1], this.segTree[2 * i + 2]);
    }

    private leafNode(i: number, val: number): void {
        let result: SegTreeNode;
        let count = new Array(this.k).fill(0);

        let rem = val % this.k;
        count[rem] = 1;
        result = { prod: rem, count };
        this.segTree[i] = result;
        return;
    }

    private mergeNodes(left: SegTreeNode, right: SegTreeNode): SegTreeNode {
        let result: SegTreeNode;
        let count = new Array(this.k).fill(0);
        let prod = (left.prod * right.prod) % this.k;

        for (let x = 0; x < this.k; x++) {
            count[x] = left.count[x];
        }

        for (let x = 0; x < this.k; x++) {
            let newRem = (left.prod * x) % this.k;
            count[newRem] += right.count[x];
        }

        result = { prod, count };
        return result;
    }


    private segTreeQuery(start: number, end: number, i: number, l: number, r: number): SegTreeNode {
        if (l >= start && r <= end) {
            return this.segTree[i];
        }

        let mid = (l + r) >> 1;

        if (end <= mid) { // entirely on the left side
            return this.segTreeQuery(start, end, 2 * i + 2, mid + 1, r);
        }

        if (start > mid) {// entirely on the right side
            return this.segTreeQuery(start, end, 2 * i + 2, mid + 1, r);
        }

        let leftNode: SegTreeNode = this.segTreeQuery(start, end, 2 * i + 1, l, mid);
        let rightNode: SegTreeNode = this.segTreeQuery(start, end, 2 * i + 2, mid + 1, r);

        return this.mergeNodes(leftNode, rightNode);
    }

    public query(start: number, end: number) {
        return this.segTreeQuery(start, end, 0, 0, this.n - 1);
    }

    private segTreeUpdate(i: number, l: number, r: number, idx: number, val: number) {
        if (l == r) {
            this.leafNode(i, val);
            return;
        }

        let mid = (l + r) >> 1;

        if (idx <= mid) {
            this.segTreeUpdate(2 * i + 1, l, mid, idx, val);
        } else {
            this.segTreeUpdate(2 * i + 2, mid + 1, r, idx, val);
        }
        this.segTree[i] = this.mergeNodes(this.segTree[2 * i + 1], this.segTree[2 * i + 2]);
    }

    public update(idx: number, val: number) {
        return this.segTreeUpdate(0, 0, this.n - 1, idx, val);
    }
}


function resultArray(nums: number[], k: number, queries: number[][]): number[] {
    let n = nums.length;
    let result = [];

    const segTree = new SegTree(n, k, nums);

    for (let q of queries) {
        let [idx, val, start, x] = q;
        segTree.update(idx, val);

        const node = segTree.query(start, n - 1);
        result.push(node.count[x]);
    }
    return result;
};